File: leetcode/2_add_two_numbers/add_two_num.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        
        if(l1==null) return l2;
        if(l2==null) return l1;
        
        ListNode head = new ListNode(0);
        ListNode res = head;
        int carry=0;
        
        while(carry!=0 || (carry==0 && l1!=null && l2!=null)){
            
            if(l1==null && l2==null){
                head.next = new ListNode(carry);
                head = head.next;
                carry=0;
            }else{
                
                int num1 = l1==null?0:l1.val;
                int num2 = l2==null?0:l2.val;
                
                int temp = (num1+num2+carry)%10;
                carry = (num1+num2+carry)/10;
                
                head.next = new ListNode(temp);
                head = head.next;
                
                if(l1!=null) l1=l1.next;
                if(l2!=null) l2=l2.next;
            }
            
        }
        head.next = l1==null ? l2:l1;
        return res.next;
        
        
    }
}

File: leetcode/2_add_two_numbers/add_two_num.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        
        ListNode* head=NULL;
        ListNode* tail=NULL;
        
        
        int carry=0;
        while(l1||l2||carry){
            int a = l1?l1->val:0;
            int b = l2?l2->val:0;
            
            int sum = a+b+carry;
            carry=0;
            if(sum>9){
                carry++;
                sum=sum-10;
            }
            ListNode* node = new ListNode(sum);
            if(!head){
                head = node;
                tail = node;
            }else{
                tail->next = node;
                tail = node;
            }
            
            l1= l1?l1->next:l1;
            l2= l2?l2->next:l2;
        }
        
        return head;
    }
};
